opal/mca/common/ucx/common_ucx_wpool.h

```diff
@@ -236,6 +246,58 @@ OPAL_DECLSPEC int opal_common_ucx_wpmem_flush(opal_common_ucx_wpmem_t *mem,
                                             int target);
 OPAL_DECLSPEC int opal_common_ucx_wpmem_fence(opal_common_ucx_wpmem_t *mem);
 
+OPAL_DECLSPEC int opal_common_ucx_flush(ucp_ep_h ep, ucp_worker_h worker,
+                                        opal_common_ucx_flush_type_t type,
+                                        opal_common_ucx_flush_scope_t scope,
+                                        ucs_status_ptr_t *req_ptr);
+
+static inline int _periodical_flush_nb(opal_common_ucx_wpmem_t *mem,
+                                       opal_common_ucx_winfo_t *winfo,
+                                       int target) {
+    int rc = OPAL_SUCCESS;
+
+    winfo->inflight_ops[target]++;
+    winfo->global_inflight_ops++;
+
+    if (OPAL_UNLIKELY(winfo->inflight_ops[target] >= MCA_COMMON_UCX_PER_TARGET_OPS_THRESHOLD) ||
+        OPAL_UNLIKELY(winfo->global_inflight_ops >= MCA_COMMON_UCX_GLOBAL_OPS_THRESHOLD)) {
+        opal_common_ucx_flush_scope_t scope;
+
+        if (winfo->inflight_req != UCS_OK) {
+            rc = opal_common_ucx_wait_request(winfo->inflight_req, winfo->worker,
+                                              "opal_common_ucx_flush_nb");
+            if(OPAL_UNLIKELY(OPAL_SUCCESS != rc)){
+                MCA_COMMON_UCX_VERBOSE(1, "opal_common_ucx_wait_request failed: %d", rc);
+                return rc;
+            }
+            winfo->inflight_req = UCS_OK;
+        }
+
+        if (winfo->inflight_ops[target] >= MCA_COMMON_UCX_PER_TARGET_OPS_THRESHOLD) {
+            scope = OPAL_COMMON_UCX_SCOPE_EP;
+            winfo->global_inflight_ops -= winfo->inflight_ops[target];
+            winfo->inflight_ops[target] = 0;
+        } else {
+            scope = OPAL_COMMON_UCX_SCOPE_WORKER;
+            winfo->global_inflight_ops = 0;
+            memset(winfo->inflight_ops, 0, winfo->comm_size * sizeof(short));
+        }
+
+        rc = opal_common_ucx_flush(winfo->endpoints[target], winfo->worker,
+                                   OPAL_COMMON_UCX_FLUSH_NB_PREFERRED, scope,
+                                   &winfo->inflight_req);
+        if(OPAL_UNLIKELY(OPAL_SUCCESS != rc)){
+            MCA_COMMON_UCX_VERBOSE(1, "opal_common_ucx_flush failed: %d", rc);
+            return rc;
+        }
+    } else if (OPAL_UNLIKELY(winfo->inflight_req != UCS_OK)) {
+        int ret;
+        do {
+            ret = ucp_worker_progress(winfo->worker);
+        } while (!ret);
```

Comment:
> 6.3.5.7 unsigned ucp_worker_progress ( ucp_worker_h worker )
> Returns
> Non-zero if any communication was progressed, zero otherwise

You should exit if ret is 0.